catalog/app/utils/PackageUri.js

```diff
@@ -0,0 +1,90 @@
+import { parse as parseQs } from 'querystring'
+import { parse as parseUrl } from 'url'
+
+import * as R from 'ramda'
+
+import { BaseError } from 'utils/error'
+
+export class PackageUriError extends BaseError {
+  // eslint-disable-next-line react/static-property-placement
+  static displayName = 'PackageUriError'
+
+  constructor(msg, uri) {
+    super(`Invalid package URI (${uri}): ${msg}`, { msg, uri })
+  }
+}
+
+function parsePackageSpec(spec, uri) {
+  if (spec.includes(':')) {
+    const [name, tag, ...rest] = spec.split(':')
+    if (!name) {
+      throw new PackageUriError(
+        '"package=" part must contain non-empty package name.',
+        uri,
+      )
+    }
+    if (!tag) {
+      throw new PackageUriError('"package=" part: tag must not be empty.', uri)
+    }
+    if (rest.length) {
+      throw new PackageUriError('"package=" part may contain only one ":".', uri)
+    }
+    return { name, tag }
+  }
+  if (spec.includes('@')) {
+    const [name, hash, ...rest] = spec.split('@')
+    if (!name) {
+      throw new PackageUriError(
+        '"package=" part must contain non-empty package name.',
+        uri,
+      )
+    }
+    if (!hash) {
+      throw new PackageUriError('"package=" part: hash must not be empty.', uri)
+    }
+    if (rest.length) {
+      throw new PackageUriError('"package=" part may contain only one "@".', uri)
+    }
+    return { name, hash }
+  }
+  return { name: spec }
+}
+
+// TODO: do we need strict parsing here (throw on extra parameters)?
+// TODO: do we need extra validation for each part (package name, path, registry, etc)?
+export function parse(uri) {
+  const url = parseUrl(uri)
+  if (url.protocol !== 'quilt+s3:') {
+    throw new PackageUriError(
+      `unsupported protocol "${url.protocol}". "quilt+s3:" is currently the only supported protocol.`,
+      uri,
+    )
+  }
+  if (!url.slashes) {
```

Comment:
`url.parse` is deprecated
https://nodejs.org/dist/latest-v14.x/docs/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost

And actual `URL` doesn't have `slashes` property

Reply:
node vs chrome (didnt test other browsers) implementations of w3c URL return different result when trying to parse our URIs, so i decided to fall back to node's 'url' module

Reply:
You can use `node` implementation of URL.

`const { URL } = require('url')`

---

I don't insist on that, but be prepared. It will shoot your leg one day